### PR TITLE
Validate binding groups

### PR DESCRIPTION
**Pull request by greggman**
Issue: #2195 

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                  buffer: {},
+                },
+              ],
+            })
+          );
+          device.createPipelineLayout({ bindGroupLayouts });
+        }, shouldError);
+      }
+    );
+  });
```
Comment (kainino0x) [inline src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts:33]: Occurs to me create\*Pipeline\* with 'auto' layout should also be able to hit this limit - if you declare and use a `@group` index that's too large, somewhere along the line it should hit this limit. I'm not sure if the spec actually says this though.

I also am not sure if the error should/would be in create\*Pipeline\* (for binding declarations used by the entry point) or createShaderModule (for any binding declaration).

Comment (kainino0x) [inline src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts:33]: setBindGroup also validates against maxBindGroups.

File: src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                binding: testValue - 1,
+                visibility: GPUShaderStage.VERTEX,
+                buffer: {},
+              },
+            ],
+          });
+        }, shouldError);
+      }
+    );
+  });
```
Comment (kainino0x) [inline src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts:28]: Similar situation with 'auto' layout, again not sure if it should be caught in creating the pipeline or the shader module

File: src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts (hunk 1 above)
Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts:33]: so createShaderModule is failing here so I added a test for that. I kind of feel like createShaderModule should not be failing though

https://github.com/gpuweb/gpuweb/issues/3815

Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts:33]: added something. PTAL

Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts:33]: actually createShaderModule is not enforcing these limits. In anycase, I'll add the test in the next push

File: src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts (hunk 2 above)
Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts:28]: added

File: src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { range } from '../../../../../common/util/util.js';
+
+import { kLimitBaseParams, makeLimitTestGroup } from './limit_utils.js';
+
+const limit = 'maxBindGroups';
+export const { g, description } = makeLimitTestGroup(limit);
+
+g.test('createBindGroupLayout,at_over')
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts]: done

File: src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts (hunk 2 above)
Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindingsPerBindGroup.spec.ts:33]: added

**Review by kainino0x — APPROVED** [review]
LGTM but thought of one more thing to test ^_^

File: src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            ? encoder.beginComputePass()
+            : encoder.beginRenderPass({
+                colorAttachments: [
+                  {
+                    view: texture.createView(),
+                    loadOp: 'clear',
+                    storeOp: 'store',
+                  },
+                ],
+              });
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts]: I added the test. I didn't see a way to use `createEncoder`.

First off it's based on gpu_test though I spent some time refactoring so `LimitTestImpl` and `GPUTest` now both share `GPUTestBase` where `GPUTestBase` is just a `Fixture` plus a `device` getter that's meant to be overriden, then `GPUTest` extends `GPUTestBase` and provides its own `device` getter.

But then, `createEncoder` calls `pass.encoder.executeBundles([rbEncoder.finish()]);` but I don't want to test `executeBundles`. I only want to test `renderBundleEncoder.finish`.

Further, I want to be able to clean up after the test but `createEncoder` provided no way to do that.

So, ended up, doing something else. We can refactor if need be.

[the remark this answers is not in the document]
Comment (kainino0x) [reply src/webgpu/api/validation/capability_checks/limits/maxBindGroups.spec.ts]: Ah... I totally forgot about the gpu_test thing. But also didn't think about the other problems.
This looks fine for now. I'm sure we could clean it up. But it's test code, so, if it's getting the job done, then it's fine.

**Pull request merged**